### PR TITLE
draft for path and path_list

### PR DESCRIPTION
**Pull request by muescha**
just to play around how to implement `--path-list`

TODO:
- [x] tests
- [x] _process to exclude 
- [ ] fix docs
- [x] add to readme (is done by `readmegen.py`)
- [ ] examples
- [x] refactor to use `pathlib`
- [ ] for windows path

### CODE REVIEW

**Comment by muescha** [comment]
this is working:
```shell
echo "/abc/def/gh.txt:/xyz/uvw/ab.app" | jc --path-list -p | pbcopy
```

```json
[
  {
    "url": "/abc/def/gh.txt",
    "scheme": null,
    "netloc": null,
    "path": "/abc/def/gh.txt",
    "parent": "/abc/def",
    "filename": "gh.txt",
    "stem": "gh",
    "extension": "txt",
    "path_list": [
      "abc",
      "def",
      "gh.txt"
    ],
    "query": null,
    "query_obj": null,
    "fragment": null,
    "username": null,
    "password": null,
    "hostname": null,
    "port": null,
    "encoded": {
      "url": "/abc/def/gh.txt",
      "scheme": null,
      "netloc": null,
      "path": "/abc/def/gh.txt",
      "parent": "/abc/def",
      "filename": "gh.txt",
      "stem": "gh",
      "extension": "txt",
      "path_list": [
        "abc",
        "def",
        "gh.txt"
      ],
      "query": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null
    },
    "decoded": {
      "url": "/abc/def/gh.txt",
      "scheme": null,
      "netloc": null,
      "path": "/abc/def/gh.txt",
      "parent": "/abc/def",
      "filename": "gh.txt",
      "stem": "gh",
      "extension": "txt",
      "path_list": [
        "abc",
        "def",
        "gh.txt"
      ],
      "query": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null
    }
  },
  {
    "url": "/xyz/uvw/ab.app",
    "scheme": null,
    "netloc": null,
    "path": "/xyz/uvw/ab.app",
    "parent": "/xyz/uvw",
    "filename": "ab.app",
    "stem": "ab",
    "extension": "app",
    "path_list": [
      "xyz",
      "uvw",
      "ab.app"
    ],
    "query": null,
    "query_obj": null,
    "fragment": null,
    "username": null,
    "password": null,
    "hostname": null,
    "port": null,
    "encoded": {
      "url": "/xyz/uvw/ab.app",
      "scheme": null,
      "netloc": null,
      "path": "/xyz/uvw/ab.app",
      "parent": "/xyz/uvw",
      "filename": "ab.app",
      "stem": "ab",
      "extension": "app",
      "path_list": [
        "xyz",
        "uvw",
        "ab.app"
      ],
      "query": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null
    },
    "decoded": {
      "url": "/xyz/uvw/ab.app",
      "scheme": null,
      "netloc": null,
      "path": "/xyz/uvw/ab.app",
      "parent": "/xyz/uvw",
      "filename": "ab.app",
      "stem": "ab",
      "extension": "app",
      "path_list": [
        "xyz",
        "uvw",
        "ab.app"
      ],
      "query": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null
    }
  }
]
```

slurp this is also working:

```shell
echo "/abc/def/gh.txt:/xyz/uvw/ab.app\n/def/hij/klm.txt:/efe/app.txt" | jc --path-list -p -s | jq ".[][] .path"
"/abc/def/gh.txt"
"/xyz/uvw/ab.app"
"/def/hij/klm.txt"
"/efe/app.txt"
```

```shell
echo "/abc/def/gh.txt:/xyz/uvw/ab.app\n/def/hij/klm.txt:/efe/app.txt" | jc --path-list -p -s
```

```json
[
  [
    {
      "url": "/abc/def/gh.txt",
      "scheme": null,
      "netloc": null,
      "path": "/abc/def/gh.txt",
      "parent": "/abc/def",
      "filename": "gh.txt",
      "stem": "gh",
      "extension": "txt",
      "path_list": [
        "abc",
        "def",
        "gh.txt"
      ],
      "query": null,
      "query_obj": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null,
      "encoded": {
        "url": "/abc/def/gh.txt",
        "scheme": null,
        "netloc": null,
        "path": "/abc/def/gh.txt",
        "parent": "/abc/def",
        "filename": "gh.txt",
        "stem": "gh",
        "extension": "txt",
        "path_list": [
          "abc",
          "def",
          "gh.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      },
      "decoded": {
        "url": "/abc/def/gh.txt",
        "scheme": null,
        "netloc": null,
        "path": "/abc/def/gh.txt",
        "parent": "/abc/def",
        "filename": "gh.txt",
        "stem": "gh",
        "extension": "txt",
        "path_list": [
          "abc",
          "def",
          "gh.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      }
    },
    {
      "url": "/xyz/uvw/ab.app",
      "scheme": null,
      "netloc": null,
      "path": "/xyz/uvw/ab.app",
      "parent": "/xyz/uvw",
      "filename": "ab.app",
      "stem": "ab",
      "extension": "app",
      "path_list": [
        "xyz",
        "uvw",
        "ab.app"
      ],
      "query": null,
      "query_obj": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null,
      "encoded": {
        "url": "/xyz/uvw/ab.app",
        "scheme": null,
        "netloc": null,
        "path": "/xyz/uvw/ab.app",
        "parent": "/xyz/uvw",
        "filename": "ab.app",
        "stem": "ab",
        "extension": "app",
        "path_list": [
          "xyz",
          "uvw",
          "ab.app"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      },
      "decoded": {
        "url": "/xyz/uvw/ab.app",
        "scheme": null,
        "netloc": null,
        "path": "/xyz/uvw/ab.app",
        "parent": "/xyz/uvw",
        "filename": "ab.app",
        "stem": "ab",
        "extension": "app",
        "path_list": [
          "xyz",
          "uvw",
          "ab.app"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      }
    }
  ],
  [
    {
      "url": "/def/hij/klm.txt",
      "scheme": null,
      "netloc": null,
      "path": "/def/hij/klm.txt",
      "parent": "/def/hij",
      "filename": "klm.txt",
      "stem": "klm",
      "extension": "txt",
      "path_list": [
        "def",
        "hij",
        "klm.txt"
      ],
      "query": null,
      "query_obj": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null,
      "encoded": {
        "url": "/def/hij/klm.txt",
        "scheme": null,
        "netloc": null,
        "path": "/def/hij/klm.txt",
        "parent": "/def/hij",
        "filename": "klm.txt",
        "stem": "klm",
        "extension": "txt",
        "path_list": [
          "def",
          "hij",
          "klm.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      },
      "decoded": {
        "url": "/def/hij/klm.txt",
        "scheme": null,
        "netloc": null,
        "path": "/def/hij/klm.txt",
        "parent": "/def/hij",
        "filename": "klm.txt",
        "stem": "klm",
        "extension": "txt",
        "path_list": [
          "def",
          "hij",
          "klm.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      }
    },
    {
      "url": "/efe/app.txt",
      "scheme": null,
      "netloc": null,
      "path": "/efe/app.txt",
      "parent": "/efe",
      "filename": "app.txt",
      "stem": "app",
      "extension": "txt",
      "path_list": [
        "efe",
        "app.txt"
      ],
      "query": null,
      "query_obj": null,
      "fragment": null,
      "username": null,
      "password": null,
      "hostname": null,
      "port": null,
      "encoded": {
        "url": "/efe/app.txt",
        "scheme": null,
        "netloc": null,
        "path": "/efe/app.txt",
        "parent": "/efe",
        "filename": "app.txt",
        "stem": "app",
        "extension": "txt",
        "path_list": [
          "efe",
          "app.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      },
      "decoded": {
        "url": "/efe/app.txt",
        "scheme": null,
        "netloc": null,
        "path": "/efe/app.txt",
        "parent": "/efe",
        "filename": "app.txt",
        "stem": "app",
        "extension": "txt",
        "path_list": [
          "efe",
          "app.txt"
        ],
        "query": null,
        "fragment": null,
        "username": null,
        "password": null,
        "hostname": null,
        "port": null
      }
    }
  ]
]

```

**Comment by muescha** [comment]
I know the `raw_output` type and the `_process` type is wrong 
maybe I don't need the `_process` - I just take it over because there is an empty `_process` in `url.py`.

**Comment by muescha** [comment]
should it makes sense to have an `path` parser with uses the url parser and then with the `_process` strip down some fields which is only needed for a path?

**Comment by muescha** [comment]
I think on windows: there must be an `;` as path delimiter and it can have an drive name like `C:\Program Files;C:\Winnt;C:\Winnt\System32`?

**Comment by muescha** [comment]
Edge Case:

- there are also paths like `./path/to/file.txt`?

```shell
echo "./path/to/file.txt" | jc --path-list -p
```
note: the `path_list` how has a `.path`:

```json
[
  {
    "path": "./path/to/file.txt",
    "parent": "path/to",
    "filename": "file.txt",
    "stem": "file",
    "extension": "txt",
    "path_list": [
      ".path",
      "to",
      "file.txt"
    ]
  }
]

```

File: jc/parsers/path_list.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+    raw_output: List[Dict] = []
+    if jc.utils.has_data(data):
+        for line in data.split(":"):
+            parsed_line = url.parse(
+                line,
+                raw=raw,
+                quiet=quiet
+            )
+            raw_output.append(parsed_line)
```
Comment (muescha) [inline jc/parsers/path_list.py:303]: suggestion from chatgpt - it is shorter but is it better readable?

```suggestion
    raw_output = [
        url.parse(line, raw=raw, quiet=quiet)
        for line in data.split(":")
        if jc.utils.has_data(data)
    ]
```

**Comment by kellyjonbrazil** [comment]
> The more I consider it, the more it seems preferable to create a `path` parser using [pathlib](https://docs.python.org/3/library/pathlib.html) rather than repurposing the `url` parser for paths. `pathlib` offers all the necessary functions for handling various edge cases effortlessly.

I agree. I'm also curious how this parser would/could work with `slurp`? If you have a list of pathlists, then they could get bundled into and array of arrays. Another approach would be to have `slurp` use `extend` rather than `append` so it stays an array of objects.

**Comment by muescha** [comment]
I anticipate that the --slurp option should generate arrays of arrays when the initial command produces an array.

I'm uncertain whether a --slurp-flat command should be created.

A workaround involves eliminating the nested array using `jq "[ .[] .[] ]"`:

```shell
echo "/abc/def/gh.txt:/xyz/uvw/ab.app\n/def/hij/klm.txt:/efe/app.txt" \
| jc --path-list -p -s | jq "[ .[] .[] ]"
```
```json
[
  {
    "path": "/abc/def/gh.txt",
    "parent": "/abc/def",
    "filename": "gh.txt",
    "stem": "gh",
    "extension": "txt",
    "path_list": [
      "abc",
      "def",
      "gh.txt"
    ]
  },
  {
    "path": "/xyz/uvw/ab.app",
    "parent": "/xyz/uvw",
    "filename": "ab.app",
    "stem": "ab",
    "extension": "app",
    "path_list": [
      "xyz",
      "uvw",
      "ab.app"
    ]
  },
  {
    "path": "/def/hij/klm.txt",
    "parent": "/def/hij",
    "filename": "klm.txt",
    "stem": "klm",
    "extension": "txt",
    "path_list": [
      "def",
      "hij",
      "klm.txt"
    ]
  },
  {
    "path": "/efe/app.txt",
    "parent": "/efe",
    "filename": "app.txt",
    "stem": "app",
    "extension": "txt",
    "path_list": [
      "efe",
      "app.txt"
    ]
  }
]
```

**Comment by muescha** [comment]
refactored to `pathlib.Path` but I have questions:

- what is the best wording? should the names be more like the python Path names? 

```python
```

**Comment by muescha** [comment]
it is [failing on windows](https://github.com/kellyjonbrazil/jc/actions/runs/7418492604/job/20186505436?pr=513) 

- should I exclude windows from the parser or should I use the PosixPath?
- I changed it to `PurePosixPath` 
- but should the parser also handle windows style path and path_list?

**Comment by kellyjonbrazil** [comment]
> 1. When using the `extension` attribute, should it return the extension with a dot (e.g., `.txt` from `path.suffix`) or without (e.g., `txt` from `url` parser)?

I would do it without the dot. Instead of using `path.suffix.strip('.')`, I would just do something like `path.suffix[1:]` since `strip` can do weird things depending on how many dots there are in the extension.

**Comment by kellyjonbrazil** [comment]
> 2. Regarding `path_list`: should it include an additional `/` as the first element ( the `url` parser do not have the `/` as the first element)?

Now that I see that I kinda like it. Maybe I should change the URL parser to match this behavior. I'm thinking I could just use the pathlib library to generate the list like you did here. I think it's good because it helps to show if it's a relative or absolute path.

**Comment by kellyjonbrazil** [comment]
> 3\. Should I extract more information from `Path`, such as `.drive` or `.root`?

Not yet - I haven't figured out why only PosixPurePath seems to pass tests cross-platform. We can always add these fields in later.

**Comment by kellyjonbrazil** [comment]
> 4\. What would be the best property names? Would it be beneficial to align them with the naming conventions used in Python's `Path` module?

I'd use the same names I used in the URL parser. I looked at the python naming convention and some made sense but I like extension better than suffix as it is more often called an extension.

**Comment by kellyjonbrazil** [comment]
> The code is currently [failing on Windows](https://github.com/kellyjonbrazil/jc/actions/runs/7418492604/job/20186505436?pr=513).
> 
> * Would it be advisable to exclude Windows from the parser, or should I opt for utilizing `PosixPath` instead?
> * I made the change to use `PurePosixPath`.
> * Additionally, should the parser be designed to accommodate Windows-style paths (with `\` delimiter) and `path_list` (with `;` delimiter)?

I'd use PurePosixPath as I did for the URL parser. I ran into the same issue and have not yet investigated why there is different behavior on different platforms. I think we can document this parser is for POSIX compliant paths only.

**Comment by kellyjonbrazil** [comment]
> I anticipate that the `--slurp` option should generate arrays of arrays when the initial command produces an array.

Makes sense - I can see how you might want to treat each array of paths separately instead of having them all in one bucket. I backed out the slurp flattening code.

**Comment by kellyjonbrazil** [comment]
Maybe we can create a `PurePath`, normalize with `path.as_posix()` and send that to the `PurePosixPath` constructor? That might allow us to support Windows paths to some extent as well, but not sure.

https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.as_posix

**Comment by kellyjonbrazil** [comment]
> * Additionally, should the parser be designed to accommodate Windows-style paths (with `\` delimiter) and `path_list` (with `;` delimiter)?

Let's hold off on the `;` delimiter for now. I think supporting Windows path lists will be difficult because there could be drive designations (e.g. `C:\WINDOWS\system`) and the colon will mess up path splitting for them. Might need to create a separate parser for Windows support.

**Comment by muescha** [comment]
> > 1. When using the `extension` attribute, should it return the extension with a dot (e.g., `.txt` from `path.suffix`) or without (e.g., `txt` from `url` parser)?
> 
> I would do it without the dot. Instead of using `path.suffix.strip('.')`, I would just do something like `path.suffix[1:]` since `strip` can do weird things depending on how many dots there are in the extension.

somehow my intention was to use `lstrip('.')` and not just `strip` :)

anyhow - I changed it to `path.suffix[1:]`

**Comment by muescha** [comment]
> *on second thought, maybe I should just keep the URL parser as-is since I believe all URL paths are considered absolute paths?

yes they are always absolute in this context 

(and somehow - not change too much an existing implementation, if some script depend on the `path_list`)

**Comment by muescha** [comment]
> > 3. Should I extract more information from `Path`, such as `.drive` or `.root`?
> 
> Not yet - I haven't figured out why only PosixPurePath seems to pass tests cross-platform. We can always add these fields in later.

ok - and on *nix the `drive` is always empty string and the `root` is always `\`
it is only interesting on windows

**Comment by muescha** [comment]
> 
> I'd use PurePosixPath as I did for the URL parser. I ran into the same issue and have not yet investigated why there is different behavior on different platforms. I think we can document this parser is for POSIX compliant paths only.

On Windows systems, paths are parsed using `Path` as a Windows path, resulting in the use of `\\` as the path delimiter instead of `/`. However, with `PurePosixPath`, the path is parsed following the conventions of *nix systems.

**Comment by muescha** [comment]
> POSIX

I did not know POSIX before - should it be named better `Unix styled path` instead of `POSIX styled path`?

**Comment by muescha** [comment]
> > I anticipate that the `--slurp` option should generate arrays of arrays when the initial command produces an array.
> 
> Makes sense - I can see how you might want to treat each array of paths separately instead of having them all in one bucket.

Also, if a parser returns a `List` instead of a `Dict`, this segment of code won't function as expected and may result in unexpected behavior.

> I backed out the slurp flattening code.

👍

**Comment by muescha** [comment]
I added some windows path checks in 66830e4
(debug prints still there to see what the windows CI produces) 

Note: On mac I need to use single quotes otherwise the by `\\x` or `\\f` or get excaped...

```shell
echo "C:\\\\Windows\\\\Program Files\\\\xfolder\\\\file.txt" | jc --path -p
echo 'C:\\Windows\\Program Files\\xfolder\\file.txt' | jc --path -p
```

**Comment by muescha** [comment]
Fun fact: `PureWindowsPath` did see an `\n` at the end on command line and on tests

**Comment by muescha** [comment]
fun fact after this PR 

```shell
find . -type f -name '*.json' | wc -l
    1000
```
🎉

**Comment by kellyjonbrazil** [comment]
Will you be ready for me to merge this soon?

**Comment by muescha** [comment]
yes it is ready to merge

**Comment by muescha** [comment]
I expect update to the docs at `README.md` and `path.md` and `path_list.md` done by `docgen.sh` and should not be included in this PR.

**Comment by kellyjonbrazil** [comment]
Looks nice, thanks!

**Pull request merged**